torch/fx/graph_module.py

```diff
@@ -296,18 +301,65 @@ def recompile(self) -> None:
         cls = type(self)
         cls.forward = _forward_from_src(self._code)
 
-        cls_call = cls.__call__
+        # If we're calling `recompile` on a GraphModule that has already
+        # been recompiled, `cls.__call__` would be `wrapped_call`.
+        # We don't want to wrap `wrapped_call` again (as this would
+        # ultimately lead to O(num_recompiles) "unwrappings" later), so
+        # we use the instance attribute `_cls_call` to store the
+        # original implementation of `__call__`
+        if not self._cls_call:
+            self._cls_call = cls.__call__
 
         def print_full_traceback(exctype, value, tb):
             traceback.print_exception(exctype, value, tb)
 
+        # Previously, if an error occurred when valid
+        # symbolically-traced code was run with an invalid input, the
+        # user would see the source of the error as coming from
+        # `File "<eval_with_key_N">`, where N is the number of times
+        # `recompile` had been called on the GraphModule. We use this
+        # function to generate a more informative error message. We
+        # return the traceback itself, a message explaining that the
+        # error occurred in a traced Module's generated forward
+        # function, and five lines of context surrounding the faulty
+        # faulty line
+        def generate_error_message(tb: TracebackType) -> str:
+            # auxiliary variables (for readability)
+            frame_summary = traceback.extract_tb(tb)[-1]
+            err_lineno = frame_summary.lineno
+            err_line_len = len(frame_summary.line)
+            all_src_lines = list(_eval_cache.values())[-1]
+
+            # constiuent substrings of the error message
+            tb_repr = traceback.format_exc()
+            custom_msg = ("Call to “exec” using an FX-traced Module, "
```

Comment:
BTW saying call to "exec" is confusing, since exec is the function that take source code and emits a code object. It would be more accurate to say call to the code object, but that's still too far into the weeds of the implementation details IMO. I think this should just say something like "Call to generated forward..."